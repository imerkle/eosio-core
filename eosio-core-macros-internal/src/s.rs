use crate::proc_macro::TokenStream;
use eosio_core::symbol_from_str;
use proc_macro2::{Literal, TokenTree};
use quote::{quote, ToTokens, TokenStreamExt};
use std::convert::TryFrom;
use syn::parse::{Parse, ParseStream, Result};
use syn::{parse_macro_input, LitInt, Token};

struct EosioSymbol(u64);

impl Parse for EosioSymbol {
    fn parse(input: ParseStream) -> Result<Self> {
        let precision = input.parse::<LitInt>()?.value();
        input.parse::<Token![,]>()?;

        let mut code = String::new();
        while !input.is_empty() {
            let segment = input.fork().parse::<TokenTree>()?.to_string();
            input.parse::<TokenTree>()?;
            code += &segment;
        }

        let precision = match u8::try_from(precision) {
            Ok(p) => p,
            Err(_) => {
                return Err(input.error(format!(
                    "precision of {} is too large; must be <= {}",
                    precision,
                    u8::max_value()
                )))
            }
        };

        symbol_from_str(precision, code.as_str())
            .map(Self)
            .map_err(|e| input.error(e))
    }
}

impl ToTokens for EosioSymbol {
    fn to_tokens(&self, tokens: &mut ::proc_macro2::TokenStream) {
        tokens.append(Literal::u64_suffixed(self.0))
    }
}

pub fn expand(input: TokenStream) -> TokenStream {
    let symbol = parse_macro_input!(input as EosioSymbol);
    quote!(#symbol).into()
}
