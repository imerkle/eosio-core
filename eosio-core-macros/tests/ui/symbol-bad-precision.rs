extern crate eosio_core_macros;

use eosio_core_macros::s;

fn main() {
    let _ = s!(256, EOS);
}