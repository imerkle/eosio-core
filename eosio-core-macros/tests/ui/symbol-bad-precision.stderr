error: unexpected end of input, precision of 256 is too large; must be <= 255
 --> $DIR/symbol-bad-precision.rs:6:13
  |
6 |     let _ = s!(256, EOS);
  |             ^^^^^^^^^^^^
  |
  = note: this error originates in a macro outside of the current crate (in Nightly builds, run with -Z external-macro-backtrace for more info)
